### PR TITLE
Extract files stored within metadata hash into ByteArrayIndexInputs

### PR DESCRIPTION
**Pull request by tlrx**
This pull request is addressing https://github.com/elastic/elasticsearch/pull/53860#discussion_r395609594 : we can now detect blobs that are stored within the metadata `hash` field and open them directly as `ByteArrayIndexInput`. There won't be stats associated to them but since they reside in memory there's no need to track their access patterns.

### CODE REVIEW

**Comment by tlrx** [comment]
@elasticmachine update branch

**Comment by tlrx** [comment]
@elasticmachine update branch

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java [hunk 1]
```diff
@@ -98,4 +85,10 @@ private void ensureValidPosition(long position) {
             throw new IllegalArgumentException("Position [" + position + "] is invalid");
         }
     }
+
+    boolean assertHashIsNotEqualToContent() {
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java]: I think it'd be better to assert this (inline) in the constructor.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 2]
```diff
@@ -221,7 +222,12 @@ public CacheFile getCacheFile(CacheKey cacheKey, long fileLength) throws Excepti
     @Override
     public IndexInput openInput(final String name, final IOContext context) throws IOException {
         ensureOpen();
+
         final BlobStoreIndexShardSnapshot.FileInfo fileInfo = fileInfo(name);
+        if (fileInfo.metadata().hashEqualsContents()) {
+            return new ByteArrayIndexInput("ByteArrayIndexInput(" + name + ')', fileInfo.metadata().hash().bytes);
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: It looks like `fileInfo.metadata().hash()` is not today a slice of the underlying byte array (i.e. it has zero `offset` and full length) but I don't think we guarantee that in future. I think we should  pass offset and length to the `ByteArrayIndexInput` constructor just to be sure.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java [hunk 3]
```diff
@@ -259,25 +257,8 @@ public String toString() {
...
-                }
-            }
-            if ((pos < 0L) || (length < 0L) || (pos + length > data.bytes.length)) {
-                throw new IllegalArgumentException("Invalid arguments (pos=" + pos + ", length=" + length
-                    + ") for hash content (length=" + data.bytes.length + ')');
-            }
-            stream = new ByteArrayInputStream(data.bytes, Math.toIntExact(pos), Math.toIntExact(length));
-        }
-        return stream;
+        assert assertHashIsNotEqualToContent();
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: If we move the assertion to the constructor then this isn't necessary.

**Review by DaveCTurner — COMMENTED** [review]
Thanks @tlrx, I left a couple of minor comments but otherwise looks good.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java]: Good point

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java (hunk 3 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: +1

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java (hunk 2 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: Sure

**Comment by tlrx** [comment]
Thanks @DaveCTurner - I addressed your comments. Can you please have another look?

**Review by DaveCTurner — APPROVED** [review]
LGTM, thanks @tlrx

**Comment by tlrx** [comment]
Thanks David!

**Pull request merged**